src/workerd/server/server.c++

```diff
@@ -1365,25 +1372,167 @@ public:
       return kj::heap<ActorChannelImpl>(*this, kj::mv(id));
     }
 
+    // ActorContainer mostly serves as a wrapper around Worker::Actor.
+    // We use it to associate a HibernationManager with the Worker::Actor, since the
+    // Worker::Actor can be destroyed during periods of prolonged inactivity.
+    //
+    // We use a RequestTracker to track strong references to this ActorContainer's Worker::Actor.
+    // Once there are no Worker::Actor's left (excluding our own), `inactive()` is triggered and we
+    // initiate the eviction of the Durable Object. If no requests arrive in the next 10 seconds,
+    // the DO is evicted, otherwise we cancel the eviction task.
+    class ActorContainer final: public RequestTracker::Hooks {
+    public:
+      ActorContainer(kj::StringPtr key, ActorNamespace& parent, kj::Timer& timer)
+          : key(key),
+            tracker(kj::refcounted<RequestTracker>(*this)),
+            parent(parent),
+            timer(timer),
+            lastAccess(timer.now()) {}
+
+      ~ActorContainer() {
+        // Shutdown the tracker so we don't use active/inactive hooks anymore.
+        tracker->shutdown();
+
+        KJ_IF_SOME(a, actor) {
+          // Unknown broken reason.
+          auto reason = 0;
+          a->shutdown(reason);
+        }
+
+        // We need to make sure we're removed from the actors map.
+        parent.onBrokenTasks.erase(key);
+        parent.actors.erase(key);
+      }
+
+      void active() override {
+        // We're handling a new request, cancel the eviction promise.
+        shutdownTask = kj::none;
+      }
+
+      void inactive() override {
+        KJ_IF_SOME(a, actor) {
+          KJ_IF_SOME(m, a->getHibernationManager()) {
+            // The hibernation manager needs to survive actor eviction and be passed to the actor
+            // constructor next time we create it.
+            manager = kj::addRef(*static_cast<HibernationManagerImpl*>(&m));
+          }
+        }
+        shutdownTask = handleShutdown();
+      }
+
+      // Processes the eviction of the Durable Object and hibernates active websockets.
+      kj::Promise<void> handleShutdown() {
+        // After 10 seconds of inactivity, we destroy the Worker::Actor and hibernate any active
+        // JS WebSockets.
+        // TODO(someday): We could make this timeout configurable to make testing less burdensome.
+        co_await timer.afterDelay(10 * kj::SECONDS);
+        KJ_IF_SOME(a, actor) {
+          if (a->isShared()) {
+            // Our ActiveRequest refcounting has broken somewhere. This is likely because we're
+            // `addRef`-ing an actor that has had an ActiveRequest attached to its kj::Own (in other
+            // words, the ActiveRequest count is less than it should be).
+            //
+            // Rather than dropping our actor and possibly ending up with split-brain,
+            // we should opt out of the deferred proxy optimization and log the error to Sentry.
```

Comment:
Note that since this is `workerd`, this won't go to Sentry.

Which makes me wonder now if this warning is even useful. But I don't think we've figured out anything better to do.

Maybe we could write the error message to be a bit clearer to users:

```
KJ_LOG(ERROR, "Detected internal bug in hibernation: Durable Object has strong references when hibernation timeout expired.");
```